[12.0][IMP] allow different workday lengths

huguesdk opened the pull request:
## description

* allow contracts with different workday lengths for different employees of the same company.
* use only the resource.calendar of the contracts to define the length of a workday, instead of using the resource.calendar of the company.
* improve compatibility with hr_holidays.
* test and fix access rights.

## odoo task

[t12627](https://gestion.coopiteasy.be/web#id=12618&model=project.task)

huguesdk committed:
[IMP] allow different workday lengths

* allow contracts with different workday lengths for different employees
  of the same company.
* use only the resource.calendar of the contracts to define the length
  of a workday, instead of using the resource.calendar of the company.
* improve compatibility with hr_holidays.
* test and fix access rights.

huguesdk committed:
[IMP] add resource.mixin.get_calendar_for_date()

carmenbianca commented on resource_work_time_from_contracts/models/resource_calendar.py:
```diff
@@ -0,0 +1,56 @@
+        # attendances that starts with the weekday of date_from then wraps
+        # around and ends with the weekday before date_from. the list should
+        # be no longer (in terms of weekdays) than date_to - date_from + 1.
+        all_attendances = self.attendance_ids
+        if not all_attendances:
```
```suggestion
        self.ensure_one()
        all_attendances = self.attendance_ids
        if not all_attendances:
```

huguesdk replied:
how do you decide whether to call `self.ensure_one()` or not? because this could be used everywhere.

carmenbianca replied:
I have frankly no idea.

carmenbianca commented on resource_work_time_from_contracts/models/resource_calendar.py:
```diff
@@ -0,0 +1,56 @@
+            return (None, None)
+        weekday_from = date_from.weekday()
+        weekday_to = date_to.weekday()
+        all_attendances = all_attendances.filtered(
+            lambda r: int(r.dayofweek) >= weekday_from
+        ) + all_attendances.filtered(lambda r: int(r.dayofweek) < weekday_from)
```
I have to assume that you meant this:

```suggestion
        all_attendances = all_attendances.filtered(
            lambda r: int(r.dayofweek) >= weekday_from
        ) + all_attendances.filtered(lambda r: int(r.dayofweek) <= weekday_to)
```

(`<= weekday_to` at the end of the last line).


To the best of my reading abilities, the current statement simplifies to `all_attendances = all_attendances`.

carmenbianca replied:
I was wrong, and this needs a comment. You're re-ordering the recordset so that it starts on `weekday_from`. I sometimes forget that Odoo sets are ordered…

carmenbianca replied:
This no longer needs to a comment because the methods were split.

carmenbianca commented on resource_work_time_from_contracts/models/resource_calendar.py:
```diff
@@ -0,0 +1,56 @@
+            first_attendance_tuple = None
+        else:
+            first_attendance_tuple = (first_attendance_date, first_attendance)
+        # do the same (but in the other direction) for date_to
```
Before the comment, add some spacing to let the method breathe. Maybe elsewhere as well.

carmenbianca replied:
No longer relevant.

carmenbianca commented on resource_work_time_from_contracts/models/resource_mixin.py:
```diff
@@ -104,40 +24,40 @@ def get_work_days_data(
         from_datetime, to_datetime = self._localize_datetimes(
             from_datetime, to_datetime
         )
+        normal_attendance_intervals = self._get_attendance_intervals(
+            from_datetime.replace(hour=0, minute=0, second=0, microsecond=0),
+            to_datetime.replace(hour=0, minute=0, second=0, microsecond=0)
+            + datetime.timedelta(days=1),
+        )
```
This uses the timezone of `from_datetime` and `to_datetime`, not the timezone of the mixin. I don't know if this is fine.

carmenbianca replied:
But: This also wouldn't be a new bug. This code is partially copied from below.

carmenbianca replied:
I was wrong.

carmenbianca commented on resource_work_time_from_contracts/models/resource_mixin.py:
```diff
@@ -203,29 +234,49 @@ def _get_work_time_per_contract(
-        return result
+        return intervals
+
+    def _get_leave_intervals(self, from_datetime, to_datetime, domain):
+        from_datetime, to_datetime = self._localize_datetimes(
+            from_datetime, to_datetime
+        )
+        # leave intervals are stored in the calendar of the resource, which
+        # should be the same as the one of the company.
+        calendar = self.resource_calendar_id
+        intervals = calendar._leave_intervals(
```
This is confusingly named. How does `_get_leave_intervals` differ from `_leave_intervals`?

carmenbianca replied:
I didn't see that `_leave_intervals()` is on the calendar. Maybe it's fine.

carmenbianca commented on resource_work_time_from_contracts/models/resource_mixin.py:
```diff
@@ -203,29 +234,49 @@ def _get_work_time_per_contract(
-    def _get_work_time_from_contracts(self, from_datetime, to_datetime, domain=None):
+    def _get_attendance_intervals(self, from_datetime, to_datetime):
         from_datetime, to_datetime = self._localize_datetimes(
             from_datetime, to_datetime
         )
         contracts = self._get_active_contracts(from_datetime.date(), to_datetime.date())
-        work_time_results = self._get_work_time_per_contract(
-            contracts, from_datetime, to_datetime, domain
+        intervals = self._get_attendance_intervals_from_contracts(
```
This is confusingly named. How does `_get_attendance_intervals` differ from `_get_attendance_intervals_from_contracts`?

carmenbianca replied:
It's probably fine.

carmenbianca commented on resource_work_time_from_contracts/models/resource_mixin.py:
```diff
@@ -142,6 +142,18 @@ def get_attendances_of_date_range(self, date_from, date_to):
             return (None, None)
         return (earliest_attendance[1], latest_attendance[1])
 
+    def get_calendar_for_date(self, date):
+        """
+        Get the resource.calendar of the contract that applies to the provided
+        date.
+        """
+        # there can be multiple contracts active at the same time. only the
+        # first one found will be used.
+        current_contracts = self._get_active_contracts(date, date)
+        if not current_contracts:
+            return None
+        return current_contracts[0].resource_calendar_id.id
```
Should this really return an integer?

Maybe it should return False or the recordset. This makes it easier to extend.

carmenbianca reviewed (changes requested):
Some comments and change requests. I'm not 100% sure I grok the implementation, but I understand the tests, so that's good.

carmenbianca committed:
[IMP] resource_work_time_from_contracts: Make more extensible

By splitting this method up, it will be easier for me to do some
extension work elsewhere.

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

huguesdk committed:
[IMP] allow to configure day rounding mode

carmenbianca committed:
[FIX] resource_work_time_from_contracts: Return record instead of integer id

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

huguesdk committed:
[FIX] handle contracts without a calendar

avoid error with contracts without a resource_calendar_id.

carmenbianca commented on resource_work_time_from_contracts/models/resource_mixin.py:
```diff
@@ -203,29 +301,49 @@ def _get_work_time_per_contract(
+        leave_intervals = self._get_leave_intervals(from_datetime, to_datetime, domain)
+        return attendance_intervals - leave_intervals
+
+    def _get_work_time_from_contracts(self, from_datetime, to_datetime):
+        """
+        Return the work time per day according to all contracts of the
+        resource, ignoring leaves.
+        """
+        intervals = self._get_attendance_intervals(from_datetime, to_datetime)
+        return self._sum_intervals(intervals)
```
should be `_get_attendance_time_from_contracts`

carmenbianca commented on resource_work_time_from_contracts/models/resource_mixin.py:
```diff
@@ -104,56 +35,209 @@ def get_work_days_data(
             work_time_per_day = self.list_normal_work_time_per_day(
-                from_datetime, to_datetime, domain=domain
+                from_datetime, to_datetime
+            )
+        return self._sum_work_days_data(work_time_per_day, normal_work_time_per_day)
+
```
The method called here is unique to this method, and should be closer.

huguesdk replied:
moved it as close as possible (while keeping private methods after public ones, and sub-methods above the ones that use them).

carmenbianca commented on resource_work_time_from_contracts/models/resource_mixin.py:
```diff
@@ -104,56 +35,209 @@ def get_work_days_data(
+                    (work_time / normal_work_time) / granularity
+                    - granularity * DAY_ROUNDING_CEIL_THRESHOLD
+                )
+                * granularity
+            )
+
+        if rounding_mode == DAY_ROUNDING_MODE_ROUND:
+            return round_day_round
+        if rounding_mode == DAY_ROUNDING_MODE_CEIL:
+            return round_day_ceil
+        return round_day_default
+
+    def _get_day_rounding_func(self):
+        ir_config_parameter_model = self.env["ir.config_parameter"].sudo()
+        rounding_mode = ir_config_parameter_model.get_param(DAY_ROUNDING_MODE_PARAM)
+        rounding_granularity = ir_config_parameter_model.get_param(
+            DAY_ROUNDING_GRANULARITY_PARAM
+        )
+        if rounding_granularity:
+            rounding_granularity = float(rounding_granularity)
+        else:
+            rounding_granularity = DEFAULT_DAY_ROUNDING_GRANULARITY
+        return self._get_round_day_round_func(rounding_mode, rounding_granularity)
+
+    def _sum_work_days_data(self, work_time_per_day, normal_work_time_per_day):
         num_days = 0.0
         num_hours = 0.0
+        round_day = self._get_day_rounding_func()
         for day, work_time in work_time_per_day:
             if work_time == 0.0:
                 continue
             normal_work_time = normal_work_time_per_day[day]
-            # we use the same rounding computation as in
-            # resource.resource_mixin.get_work_days_data().
-            num_days += (
-                float_utils.round(ROUNDING_FACTOR * work_time / normal_work_time)
-                / ROUNDING_FACTOR
-            )
+            num_days += round_day(work_time, normal_work_time)
             num_hours += work_time
```
Are we certain that this should be continuously rounded in the loop, instead of rounded once at the end?

huguesdk replied:
yes. this is how it’s done in the `resource` module too. it’s the value for each day that must be rounded on its own, and then all of these summed together. this is especially true when using the non-default rounding methods, where the value of each day is rounded according to a defined granularity.

carmenbianca commented on resource_work_time_from_contracts/models/resource_mixin.py:
```diff
@@ -104,56 +35,209 @@ def get_work_days_data(
+            return None
+        return current_contracts[0].resource_calendar_id
+
+    def _sum_intervals(self, intervals):
+        result = defaultdict(float)
+        for start, stop, meta in intervals:
+            result[start.date()] += (stop - start).total_seconds() / 3600
+        return sorted(result.items())
```
Bad method name. `_hours_per_date[_from_intervals]` would work better.

carmenbianca replied:
In fact, `_hours_per_day_from_intervals` already exists… It appears to do something frustratingly adjacent, but subtly different.

huguesdk replied:
renamed `sum_intervals()` to `_sum_hours_per_date_from_intervals()` and `_hours_per_day_from_intervals()` to `_expected_attendance_per_day_from_intervals()`.

carmenbianca commented on resource_work_time_from_contracts/models/resource_mixin.py:
```diff
@@ -104,56 +35,209 @@ def get_work_days_data(
+        return round_day_default
+
+    def _get_day_rounding_func(self):
+        ir_config_parameter_model = self.env["ir.config_parameter"].sudo()
+        rounding_mode = ir_config_parameter_model.get_param(DAY_ROUNDING_MODE_PARAM)
+        rounding_granularity = ir_config_parameter_model.get_param(
+            DAY_ROUNDING_GRANULARITY_PARAM
+        )
```
Do these come with defaults? We should either fall back to a default, or populate these params by default, or both.

huguesdk replied:
yes, all of these has defaults. `ir.config_parameter.get_param()` return `False` if the parameter is not defined. in that case, we use:

*   `DAY_ROUNDING_MODE_PARAM`: `round_day_default()`
*   `DAY_ROUNDING_GRANULARITY_PARAM`: `DEFAULT_DAY_ROUNDING_GRANULARITY` (`1`)

carmenbianca commented on resource_work_time_from_contracts/models/resource_mixin.py:
```diff
@@ -172,14 +256,28 @@ def _get_active_contracts(self, date_start, date_end):
-        self, contracts, from_datetime, to_datetime, domain
+    def _get_attendance_intervals_of_calendar(
+        self, from_datetime, to_datetime, calendar
     ):
         """
-        Return the work time per day per contract.
+        Return the attendance intervals for the provided resource.calendar,
+        ignoring leaves.
         """
-        work_time_results = []
+        return calendar._attendance_intervals(
+            from_datetime, to_datetime, self.resource_id
+        )
```
Better to get rid of this. There's already many tiny methods, and this one seems superfluous.

huguesdk replied:
i prefer to leave it there. it’s because these kind of methods are not used in the `resource` module that we had to override so many things.

carmenbianca reviewed (changes requested):
Did a second pass. Some comments.

huguesdk committed:
[IMP] improve naming of methods

carmenbianca reviewed (approved):
LGTM, amazing work :)

huguesdk commented:
/ocabot merge major
